File: rabbit/consumer.js
var amqp = require('amqplib');
var config = JSON.parse(process.argv[2]);

var msgsToSend = config.msgsToSend;

amqp.connect('amqp://localhost').then(function(conn) {
  process.once('SIGINT', function() { conn.close(); });
  return conn.createChannel().then(function(ch) {
    console.log('time (ms),latency (ms),consume (ms)');
    var ok = ch.assertQueue(config.queue, {autoDelete: true, durable: config.queueDurable})
    .then(function() {
      return ch.purgeQueue(config.queue);
    })
    .then(function() {
      return ch.prefetch(config.prefetchCount);
    })
    .then(function() {
      var start;
      var last;

      return ch.consume(config.queue, function(msg) {
        var now = Date.now();
        var then = 1*msg.content.toString();
        if (last === undefined) {
          start = now;
          last = now;
        }
        console.log((now-start)+','+(now-then)+','+(now-last));
        last = now;

        if (config.msgAckDelay > 0) {
          setTimeout(ack, config.msgAckDelay);
        } else {
          ack();
        }

        function ack() {
          if (! config.autoAck) {
            ch.ack(msg);
          }
          msgsToSend--;
          if (msgsToSend == 0) {
            setImmediate(function() {conn.close();});
          }
        }
      }, {noAck: config.autoAck});
    });
  });
}).then(null, , function(err) {
  console.warn('producer err', err);
  process.exit(1);
});

File: rabbit/java/src/main/java/Consumer.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.QueueingConsumer;

public class Consumer {

  public static void main(String[] argv) throws Exception {
    Map<String, String> params = new HashMap<String, String>();
    Arrays.asList(argv).stream()
    .forEach(s -> { String[] ss = s.split("="); params.put(ss[0], ss[1]); } );
    String queueName = params.get("queue");
    int msgsToSend = Integer.parseInt(params.get("msgsToSend"));
    int msgAckDelay = Integer.parseInt(params.get("msgAckDelay"));
    int prefetchCount = Integer.parseInt(params.get("prefetchCount"));
    boolean autoAck = Boolean.parseBoolean(params.get("autoAck"));

    System.out.println("ts (ms),time (ms),latency (ms),consume (ms)");

    ConnectionFactory factory = new ConnectionFactory();
    factory.setHost("localhost");
    Connection connection = factory.newConnection();
    Channel channel = connection.createChannel();
    channel.basicQos(prefetchCount);

    channel.queueDeclare(queueName, false, false, true, null);

    QueueingConsumer consumer = new QueueingConsumer(channel);
    channel.basicConsume(queueName, autoAck, consumer);

    long last = -1;
    long start = 0;
    for (int i = 0; i < msgsToSend; i++) {
      QueueingConsumer.Delivery delivery = consumer.nextDelivery();
      long now = System.currentTimeMillis();

      long then = Long.parseLong(new String(delivery.getBody()));
      if (last == -1) {
    	  start = now;
    	  last = now;
      }
      System.out.println(now+","+(now-start)+","+(now-then)+","+(now-last));
      last = now;

      if (msgAckDelay > 0) {
    	  Thread.sleep(msgAckDelay);
      }

      if (!autoAck) {
    	  channel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
      }
    }
    channel.close();
    connection.close();
  }
}
